### PR TITLE
Close search contexts on reassigned shard

### PR DESCRIPTION
**Pull request by henningandersen**
If a shard is reassigned to a node, but it has open searches (could be
scrolls even), the current behavior is to throw a
ShardLockObtainFailedException. This commit changes the behavior to
close the search contexts, likely failing some of the searches. The
sentiment is to prefer restoring availability over trying to complete
those searches. A situation where this can happen is when master(s) are
restarted, which is likely to cause similar search issues anyway.

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.elasticsearch.transport.TransportService;
+import org.hamcrest.Matchers;
+
+import java.util.Collection;
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 0, scope= ESIntegTestCase.Scope.TEST)
```
Comment (aimiebell) [inline server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java:30]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

Comment (henningandersen) [reply server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java]: @aimiebell thanks for your interest in this pull request. I fail to spot where this does not adhere to normal java conventions, can you please point me to the specific line that caught your interest and make a suggestion for a change?

File: server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        });
+    }
+
+    private void testLostMaster(CheckedBiConsumer<String, String, Exception> loseMaster) throws Exception {
+        final String master = internalCluster().startMasterOnlyNode();
+        final String dataNode = internalCluster().startDataOnlyNode();
+
+        index("test", "test", "{}");
+
+        assertThat(client().prepareSearch("test").setScroll("1m").get().getScrollId(), is(notNullValue()));
```
Comment (ywelsch) [inline server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java:71]: perhaps set this much larger than "1m"?

File: server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java [hunk 3]
```diff
@@ -108,6 +108,14 @@ default void afterIndexRemoved(Index index, IndexSettings indexSettings, IndexRe
 
     }
 
+    /**
+     * Called during shad create before obtaining the shard lock.
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java:112]: s/shad/shard

File: server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java [hunk 4]
```diff
@@ -108,6 +108,14 @@ default void afterIndexRemoved(Index index, IndexSettings indexSettings, IndexRe
...
+     * @param routing the routing entry that caused the shard to be created.
+     * @param indexSettings the shards index settings
+     */
+    default void beforeShardLockDuringShardCreate(ShardRouting routing, Settings indexSettings) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java:116]: can we repurpose the `beforeIndexShardCreated` method? Does not look to be seriously used anywhere

**Review by ywelsch — APPROVED** [review]
I like the approach taken here. I've left some minor comments, looking good o.w.

File: server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java (hunk 2 above)
Comment (henningandersen) [reply server/src/internalClusterTest/java/org/elasticsearch/search/SearchServiceCleanupOnLostMasterIT.java]: Made it 30m in ea3631ee4d29e087c5f11b441f0448af0d3cd354

File: server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java (hunk 4 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java]: Yes, thanks, that makes the naming much nicer, done in d8695264bfdf458a47c085198ba08449943ce523

**Pull request merged**